src/transformStateWithClones.js

```diff
@@ -6,8 +6,48 @@
  *
  * @return {Object[]}
  */
+
+const ACTION_TYPES = {
+  add: 'addProperties',
+  remove: 'removeProperties',
+  clear: 'clear',
+};
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+
+  for (const { type, extraData, keysToRemove } of actions) {
+    const stateCopy = { ...(history[history.length - 1] ?? state) };
+    let historyItem = {};
```

Comment:
it's unnecessary variable  
```suggestion
```

Reply:
AI said: `Pushing the state copy to the history array should be done after the switch block to avoid repetition.`
So I just move my value to the historyItem variable. Which solution could be better?